open-craft/xblock-vectordraw
# Initial implementation

## VectorDraw XBlock

This XBlock allows course authors to define vector drawing exercises. To complete a vector drawing exercise, students must correctly place a set of predefined vectors (and points) on a board.

All data associated with this XBlock is stored in XBlock fields. Vectors, points, and expected result for an exercise must be defined using JSON, but we are also planning on adding WYSIWYG functionality to Studio that will allow authors to specify vectors and expected results by drawing them (cf. below). *This functionality will not replace existing functionality for specifying elements that are part of the exercise; it will instead be offered as an alternative method for creating content.* For a given exercise, student state consists of:

- the set of vectors and points that were present on the board when student last submitted an answer
- a list of checks to perform in order to obtain grade
- result returned by the grader for the most recent answer (which contains info about correctness of answer and feedback message).

The amount of data to store for an exercise depends on how complex it is (i.e., how many vectors and points there are in total); based on [examples](https://github.com/open-craft/jsinput-vectordraw/tree/master/Notes_and_Examples) from an existing JSInput implementation, it should not exceed 1-2K per problem, per student. Grading happens by comparing the set of elements on the board to a set of conditions specified by the course author ("expected result"). Minimally, this involves checking for each required element whether or not it is present in a given answer. Additional checks may include:

- checking coordinates of points
- checking coordinates of tail and/or tip of a vector
- checking the length of a vector
- checking the angle of a vector
- checking whether a line passes through one or more points.

None of these operations are very expensive computationally, and the grader short-circuits as soon as it encounters a failing check (there is currently no notion of a partially correct answer), so we don't expect there to be any performance bottlenecks. We haven't performed any load testing.

We'd like to get this XBlock on edx.org for Davidson College by mid-December.

**Testing**

1. Install the XBlock on Devstack via:

        pip install -r requirements.txt

2. Add `vectordraw` to the Advanced Module List in your course settings.
3. Add a "Vector Drawing" block to a unit.
4. Customize the Vector Drawing block. You can use one of the examples from the jsinput-vectordraw repository (e.g., [2_boxIncline_multiVector](https://raw.githubusercontent.com/open-craft/jsinput-vectordraw/master/Notes_and_Examples/2_boxIncline_multiVector/2_boxIncline.html)) for this purpose. Note that values of "Vectors", "Points", "Expected result" settings need to be valid JSON.

---

**Partner information** hosted on edx.org (Davidson College)
**Merge deadline** mid-December
**JIRA Story** [OSPR-935](https://openedx.atlassian.net/browse/OSPR-935)
**Confluence / Product Asset** N/A
**Sandbox URL** [LMS](http://pr10515.sandbox.opencraft.com/) ([Example](http://pr10515.sandbox.opencraft.com/courses/course-v1:OpenCraft+VD101+2015/courseware/be8fac39945d427c987af0bdbec200c3/ecb2b2ba39214ee1b272dd83e8ad2890/)), [Studio](http://studio.pr10515.sandbox.opencraft.com/) ([Example](http://studio.pr10515.sandbox.opencraft.com/container/block-v1:OpenCraft+VD101+2015+type@vertical+block@2ecf812afd434f359480c7852ce12fa0); see author notes below about a known glitch)
**Dependencies** N/A

---

**PR Author(s) Notes / To-Do**
* [x] Moving controls for interacting with the board above the active board (as per [this request](https://github.com/open-craft/xblock-vectordraw/pull/1#issuecomment-153837342)).
* [x] There's a glitch in Studio that keeps vectors from being rendered properly when first loading the page (you have to click "Reset" to get a proper preview). We'll have another opportunity to address this in the context of the next item.
* [x] WYSIWYG functionality for defining vectors and expected results

---

**Reviewers**
- [x] Code: @smarnach
- [x] Code: @nedbat
- [x] Code: @singingwolfboy
- [ ] Accessibility: @cptvitamin
